client/state/reader/posts/actions.js

```diff
@@ -49,93 +27,69 @@ function fetchForKey( postKey ) {
  *
  * @param  {Array}  posts Posts received
  * @return {Object} Action object
  */
-export function receivePosts( posts ) {
-	return function( dispatch ) {
-		if ( ! posts ) {
-			return Promise.resolve( [] );
-		}
-
-		const withoutUndefined = reject( posts, isUndefined );
-		const normalizedPosts = map( withoutUndefined, runFastRules );
-
-		const [ postsToReload, postsToProcess ] = partition( normalizedPosts, '_should_reload' );
-		forEach( postsToReload, post => {
-			delete post._should_reload;
-			dispatch( reloadPost( post ) );
-		} );
+export const receivePosts = posts => dispatch => {
+	if ( ! posts ) {
+		return Promise.resolve( [] );
+	}
 
-		forEach( map( postsToProcess, runSlowRules ), slowPromise => {
-			slowPromise.then( post => {
-				dispatch( {
-					type: READER_POSTS_RECEIVE,
-					posts: [ post ],
-				} );
+	const normalizedPosts = compact( posts ).map( runFastRules );
 
-				// keep the old feed post store in sync
-				Dispatcher.handleServerAction( {
-					data: post,
-					type: action.RECEIVE_NORMALIZED_FEED_POST,
-				} );
-			} );
-		} );
+	// save the posts after running the fast rules
+	dispatch( {
+		type: READER_POSTS_RECEIVE,
+		posts: normalizedPosts,
+	} );
 
+	// also save them after running the slow rules
+	Promise.all( normalizedPosts.map( runSlowRules ) ).then( processedPosts => {
 		dispatch( {
 			type: READER_POSTS_RECEIVE,
-			posts: normalizedPosts,
+			posts: [ processedPosts ],
 		} );
+	} );
 
-		// keep the old feed post store in sync
-		forEach( normalizedPosts, post => {
-			const postForFlux = has( post, '_should_reload' ) ? omit( post, '_should_reload' ) : post;
-			Dispatcher.handleServerAction( {
-				data: postForFlux,
-				type: action.RECEIVE_NORMALIZED_FEED_POST,
-			} );
-		} );
+	forEach( filter( normalizedPosts, 'railcar' ), trackRailcarRender );
+
+	// TODO: resolve weird dependency between related-posts and the return here
+	return Promise.resolve( normalizedPosts );
+};
 
-		forEach( filter( postsToProcess, 'railcar' ), trackRailcarRender );
+export const fetchPost = postKey => dispatch => {
+	return fetchForKey( postKey )
+		.then( data => dispatch( receivePosts( [ data ] ) ) )
+		.catch( error => dispatch( receiveErrorForPostKey( error, postKey ) ) );
+};
 
-		return Promise.resolve( normalizedPosts );
+export function reloadPost( post ) {
+	return dispatch => {
+		// keep track of any railcars we might have
+		const railcar = post.railcar;
+		return fetchForKey( keyForPost( post ) ).then( data => {
+			data.railcar = railcar;
+			return dispatch( receivePosts( [ data ] ) );
+		} );
+	};
+}
+
+function receiveErrorForPostKey( error, postKey ) {
+	return {
+		type: READER_POSTS_RECEIVE,
+		posts: [
+			{
+				feed_ID: postKey.feedId,
+				ID: postKey.postId,
+				site_ID: postKey.blogId,
+				is_external: ! postKey.blogId,
+				global_ID: `${ postKey.feedId || 'na' }-${ postKey.blogId || 'na' }-${ postKey.postId }`,
```

Comment:
this is inconsistent with the `uuid` earlier... do we need both?

Reply:
i should probably switch this one to a uuid as well